### ISSUE TITLE
lxcfs[631]: *** Error in `/usr/bin/lxcfs': double free or corruption (fasttop): 0x75a00698 ***

### ISSUE THREAD

**Issue by nadalle** [lxc/lxcfs#173]
I can trivially repro this lxcfs memory corruption with the following command:

``while true; do echo -n `cat /proc/stat | grep -c btime`; done``
`111111111111111111111111111[snipped]11111111111111111111111111^C`
`root@test:~# error: not found`
`user@host-os ~ $`

``lxcfs[631]: *** Error in `/usr/bin/lxcfs': double free or corruption (fasttop): 0x75a00698 ***``
`systemd[1]: lxcfs.service: Main process exited, code=killed, status=6/ABRT`

$ lxcfs --version
2.0.6

OS: Ubuntu 16.04.1 LTS + lxd stable xenial PPA (I installed the PPA because the regular lxd was also unstable)
Platform: armhf
`Linux 3.14.79-cubox #4 SMP Sun Oct 23 16:23:48 CEST 2016 armv7l armv7l armv7l GNU/Linux`

**Comment by hallyn** [comment]
how long do you have to let it run for this to happen?

I'm running 2.0.6-1~ubuntu16.04.1~ppa1 but haven't reproduced this.

**Comment by nadalle** [comment]
I'm attempting to nail the repro down better now. It reproduced very easily earlier, but seems to have heisenberged itself now. It's possible the repro script was a red herring.

In general, lxcfs seems to reliably crash with memory corruption on my system within a few hours. I've been trying to run this repro (or find a better one) for a few hours now with gdb attached to lxcfs, but no luck.

There has been some other log spam though:

lxcfs[2847]: bindings.c: 2596: do_write_pids: Error writing pid to child: Bad file descriptor.
lxcfs[2847]: bindings.c: 2406: pid_from_ns: Timeout reading from parent.
lxcfs[2847]: bindings.c: 809: cgfs_iterate_cgroup: Failed closedir for /: Bad file descriptor
lxcfs[2847]: bindings.c: 3223: proc_meminfo_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3523: proc_stat_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3550: proc_stat_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3223: proc_meminfo_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3523: proc_stat_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3223: proc_meminfo_read: Internal error: truncated write to cache.
lxcfs[2847]: bindings.c: 3523: proc_stat_read: Internal error: truncated write to cache.

**Comment by nadalle** [comment]
Some more spam for your enjoyment, this time during an attempt to lxc stop --force which is hanging forever, apparently due to lxcfs:

kernel: [ 2880.289552] INFO: task (agetty):14735 blocked for more than 120 seconds.
kernel: [ 2880.311521]       Not tainted 3.14.79-cubox #4
kernel: [ 2880.333099] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
kernel: [ 2880.354694] (agetty)        D 807126e8     0 14735   1993 0x00000801
kernel: [ 2880.354779] [<807126e8>] (__schedule) from [<80221d50>] (__fuse_request_send+0xf0/0x2ec)
kernel: [ 2880.354839] [<80221d50>] (__fuse_request_send) from [<802261d8>] (fuse_dentry_revalidate+0x160/0x2d8)
kernel: [ 2880.354868] [<802261d8>] (fuse_dentry_revalidate) from [<80110df4>] (lookup_fast+0x234/0x290)
kernel: [ 2880.354894] [<80110df4>] (lookup_fast) from [<80112be8>] (link_path_walk+0x17c/0x8b8)
kernel: [ 2880.354917] [<80112be8>] (link_path_walk) from [<8011545c>] (path_openat+0x8c/0x688)
kernel: [ 2880.354970] [<8011545c>] (path_openat) from [<80116804>] (do_filp_open+0x2c/0x88)
kernel: [ 2880.355002] [<80116804>] (do_filp_open) from [<80106d80>] (do_sys_open+0x104/0x1c8)
kernel: [ 2880.355035] [<80106d80>] (do_sys_open) from [<8000e428>] (__sys_trace_return+0x0/0x18)

**Comment by Blub** [comment]
If attaching with gdb "avoids" the issue, you could instead enable core-dumps and then load that with gdb once it has crashed? Because a backtrace of lxcfs itself would definitely be useful.

**Comment by nadalle** [comment]
I'm attempting that now. I've been running into alternate failure modes where my container processes get stuck in D state waiting for lxcfs so far (I tried attaching to lxcfs and didn't see much going on).

**Comment by nadalle** [comment]
Ok, I'm currently unable to reproduce the memory corruption assert, because the various test scripts I've tried (just running various commands reading from /proc etc., eg. while true; do ps aux; done) get locked up in D state on the lxcfs fuse mount after a while, and everything grinds to a halt.

I manually core dumped the lxcfs binary in this state using a SIGABRT, but the threads all appear quiescent: [gdb.txt](https://github.com/lxc/lxcfs/files/761610/gdb.txt)

**Comment by nadalle** [comment]
As a side note, I'm also consistently unable to stop the containers using lxc stop or lxc stop --force, and have to kill -9  the lxd forkstart process associated with the container instead.

**Comment by hallyn** [comment]
Ok, there are several ways I could have been stupid that would only manifest themselves on arm.  Looks likely to be something like that.  @stgraber or @brauner or @Blub do you have a armhf available that you can test on?

**Comment by nadalle** [comment]
Ok, I managed to capture a core dump. It didn't seem to be associated with any of my repro activities, it just happened randomly.

Unfortunately, there don't seem to be any debugging symbols for lxcfs available, so the gdb info isn't that great: 
[gdb_lxcfs.txt](https://github.com/lxc/lxcfs/files/762364/gdb_lxcfs.txt)

**Comment by stgraber** [comment]
@hallyn I've got armhf yeah, been running the reproducer for a few minutes now.

    root@delmak:~# uname -a
    Linux delmak 4.4.0-62-generic #83-Ubuntu SMP Wed Jan 18 14:10:54 UTC 2017 armv7l armv7l armv7l GNU/Linux

So that's running on the official Ubuntu generic armhf kernel instead of @nadalle's custom kernel, not sure if that makes a difference. My system is a 4 core imx6 with 2GB of RAM, not too dissimilar from some of the cubox configs I could find.

**Comment by nadalle** [comment]
@stgraber FWIW I didn't have any luck reproducing the double free with the stressor, but I was pretty easily able to get guest processes to lock up reading from FUSE by just applying simple load (e.g. ps aux in a loop).

The double free seems to reliably happen when I'm not looking. Perhaps a periodic process?

**Comment by nadalle** [comment]
Any possibility of finding debug symbols for this deb somewhere?

Version: 2.0.6-1~ubuntu16.04.1~ppa1

sha1sum: a92306d5aeaeb8659ef2ae949d23495cfc52769c  lxcfs_2.0.6-1~ubuntu16.04.1~ppa1_armhf.deb

**Comment by stgraber** [comment]
If you're using ppa:ubuntu-lxc/daily, then you can go to:

https://launchpad.net/~ubuntu-lxc/+archive/ubuntu/daily/+sourcepub/7443054/+listing-archive-extra which lists all the artifacts for that build, including the ddeb packages which include the debug symbols.

If you're using one of the other PPA, then you should also install the .deb from the link above, otherwise even though both packages will have a matching version number they may slightly differ as they'll have been separate builds.

**Comment by nadalle** [comment]
Thanks! I was able to track it down and verify.

Here's some more useful core dump info. Hope we got lucky (given it's heap corruption) and there's a clue: [gdb_lxcfs_full_syms.txt](https://github.com/lxc/lxcfs/files/766004/gdb_lxcfs_full_syms.txt)

**Comment by nadalle** [comment]
I've upgraded this machine to linux 4.9.12. I'll add more information if it still reproduces (or if it doesn't).

**Comment by nadalle** [comment]
Unable to reproduce with 4.9.12.

**Comment by hallyn** [comment]
On Sat, Mar 04, 2017 at 06:48:52PM -0800, Elladan wrote:
> Unable to reproduce with 4.9.12.

Fascinating.

What could the kernel be doing...   Just to be sure, if you boot an
older kernel it still happens?  I.e. it's not an inadvertant glibc
change which came along with the kernel?

**Comment by nadalle** [comment]
On Mar 7, 2017 6:21 AM, "Serge Hallyn" <notifications@github.com> wrote:

On Sat, Mar 04, 2017 at 06:48:52PM -0800, Elladan wrote:
> Unable to reproduce with 4.9.12.

Fascinating.

What could the kernel be doing... Just to be sure, if you boot an
older kernel it still happens? I.e. it's not an inadvertant glibc
change which came along with the kernel?


The kernel was the only changed package (I found a newer armbian kernel for
cubox).

Well, I guess there was another configuration change: I swapped micro SD
cards, and the new one is much faster. I notice lxcfs uses a lot of threads
so it might affect timing somehow.

I actually did that during the time I was looking for a repro with 4.9, and
didn't see the error with the old card either, but there was only about 1
day of runtime.
